src/main/java/pl/futurecollars/invoicing/db/file/FileBasedDatabase.java

```diff
@@ -0,0 +1,98 @@
+package pl.futurecollars.invoicing.db.file;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import pl.futurecollars.invoicing.db.Database;
+import pl.futurecollars.invoicing.model.Invoice;
+import pl.futurecollars.invoicing.utils.FilesService;
+import pl.futurecollars.invoicing.utils.JsonService;
+
+@AllArgsConstructor
+public class FileBasedDatabase implements Database {
+
+  private final Path databasePath;
+  private final IdService idService;
+  private final FilesService filesService;
+  private final JsonService jsonService;
+
+  @Override
+  public int save(Invoice invoice) {
+    try {
+      invoice.setId(idService.getNextIdAndIncreament());
+      filesService.appendLineToFile(databasePath, jsonService.toJson(invoice));
+
+      return invoice.getId();
+    } catch (IOException ex) {
+      throw new RuntimeException("Database failed to save invoice", ex);
+    }
+  }
+
+  @Override
+  public Optional<Invoice> getById(int id) {
+    try {
+      return filesService.readAllLines(databasePath)
+          .stream()
+          .filter(line -> containsId(line, id))
+          .map(line -> jsonService.toObject(line, Invoice.class))
+          .findFirst();
+    } catch (IOException ex) {
+      throw new RuntimeException("Database failed to get invoice with id: " + id, ex);
+    }
+  }
+
+  @Override
+  public List<Invoice> getAll() {
+    try {
+      return filesService.readAllLines(databasePath)
+          .stream()
+          .map(line -> jsonService.toObject(line, Invoice.class))
+          .collect(Collectors.toList());
+    } catch (IOException ex) {
+      throw new RuntimeException("Failed to read invoices from file", ex);
+    }
+  }
+
+  @Override
+  public void update(int id, Invoice invoiceForUpdating) {
+    if (getById(id).isPresent()) {
+
+      invoiceForUpdating.setId(id);
+
+      List<Invoice> updatedInvoices = getAll()
```

Comment:
Przy obecnej implementacji tej metody masz dwie listy, identycznej długości, której wszystkie elementy (po za jednym) są identyczne. A więc większość danych, które masz w tej liście jest zduplikowana.

Spróbuj operować tylko na jednej liście (tej zwracanej przez getAll()).
Znajdź na niej element, który chcesz zupdate'ować, usuń go z listy zwróconej przez getAll(), zdeserializuj go, ustaw mu pola na odpowiednie wartości, zdeserializuj, dorzuć do liste zwróconej przez getAll() i zapisz tę listę do pliku.

Reply:
A w zasadzie to masz tutaj 3 razy niemal tę samą listę, ponieważ metoda getById tworzy taką listę raz jeszcze.